Multirow insert

ashudeep-sharma opened the pull request:
This Pull Request contains the changes for supporting MultiRowInsert changes on the VTGate side.
@sougou : PTAL.

sougou commented on go/sqltypes/result.go:
```diff
@@ -117,3 +120,45 @@ func (result *Result) StripFieldNames() *Result {
+// AppendResult will combine the Results Objects of one result
+// to another result.Note currently it doesn't handle cases like
+// if two results have different fields.We will enhance this function.
+func (result *Result) AppendResult(qr, innerqr *Result) {
```
This should be `func (result *Result) Append(src *Result)`, where `src` is appended to `result`. The function name doesn't need to contain Result because it's implicit.

sougou commented on go/vt/vtgate/router.go:
```diff
@@ -281,7 +285,8 @@ func (rtr *Router) execUpdateEqual(vcursor *requestContext, route *engine.Route)
 	}
-	rewritten := sqlannotation.AddKeyspaceID(route.Query, ksid, vcursor.comments)
+	keyspaceIds = append(keyspaceIds, ksid)
+	rewritten := sqlannotation.AddKeyspaceIDs(route.Query, keyspaceIds, vcursor.comments)
```
This can be simplified to: `rewritten := sqlannotation.AddKeyspaceIDs(route.Query, [][]byte{ksid}, vcursor.comments)`. Same comments below.

sougou commented on go/vt/vtgate/router.go:
```diff
@@ -357,47 +364,35 @@ func (rtr *Router) execInsertUnsharded(vcursor *requestContext, route *engine.Ro
 	if err != nil {
 		return nil, fmt.Errorf("execInsertSharded: %v", err)
 	}
+	keyspace, routing, shardKsidMap, err := rtr.getInsertShardedRoute(vcursor, route)
```
I think this can be simplified by combining it with rewriteMultiRowQuery, where this one step returns the `Mid` strings instead of routing. We could actually push this one step further to directly return shardQueries (instead of routing & shardKsidMap). We'll need to look at the code to see if that would make sense.

sougou commented on go/vt/vtgate/scatter_conn.go:
```diff
@@ -187,16 +183,63 @@ func (stc *ScatterConn) ExecuteMulti(
-		return nil, err
+// ExecuteMultiShard is like ExecuteMulti,
+// but each shard gets its own Sql Queries and BindVariables.
+func (stc *ScatterConn) ExecuteMultiShard(
```
As discussed, merge this with ExecuteMulti.

ashudeep-sharma commented:
I have incorporated all the review comments apart from getInsertShardedRoute() one, will checkout the feasability on how it can be done in a better way.
Also as discussed will be adding an end to end test-case in vtgatev3_test.py

ashudeep-sharma commented:
- Refactored the code for getInsertShardedRoute() to directly return shardQueries.

- Added test-cases in vtgatev3_test.py for MultiRow Insert as well some tests for AddKeyspaceIDs()

@sougou :PTAL and let me know in case you need more tests.

sougou commented:
LGTM
